tlvf: CmduMessageTx ->TlvList refactor

tomereli opened the pull request:
Currently, CmduMessage (and its derived classes CmduMessageTx & CmduMessageRx) include the TLV class vector and the buffer.

Some TLVs, such as vendor-specific and WSC, may contain arbitrary data that can be parsed/created with the same mechanics that we use in the CmduMessage and are required by multiple features (#471, #160, #318).

It would be easier to implement support for such TLVs using a TLV list class, which will include all the buffer handling that is currently part of the CmduMessage classes.

This PR refactors TLVF base files and adds a new class, `TlvList`, which can be used to create and parse TLVs with arbitrary data represented with an internal TLV list.

It also adds support for the first outer TLV, the vendor-specific TLV, by adding an unknown length buffer to the end of its YAML file, which is used as the buffer for the inner TLV list.
The inner TLV list is the `beerocks_header` which derives from `TlvList` and adds some backward compatible APIs so the change will not be too big to swallow in a single PR (Nevertheless, this kind of change is pretty big since in order for not breaking the build in each commit).

Now, each vendor-specific CMDU is built using a vendor-specific TLV with a payload buffer (`uint8_t`).
The buffer size is preallocated in`add_intel_vs_data()` to the maximum allowed TLV size - which is `MTU - CMDU header length - EOM TLV length - TS TLV init length` for standard CMDUs since the ieee1905 transport performs fragmentation on TLV boundary and the maximum buffer size for CMDUs containing oversized TLVs (referred to as "jumbo" CMDUs) which are not meant to be sent externally.
Having a preallocated big enough buffer allows the internal TLV list to accommodate variable-length data in the internal classes, which is mandatory since by the time we are allocating the buffer we do not know if the caller will need more space.

This brings the need to resize the inner TLVs and update the outer TLVs length on `finalize()`, which does so by going through the CMDU's inner tlv map, resizing each inner TLVlist to its final message length and moving the rest of the CMDU buffer back to the new end of the inner TLV list.

tomereli commented:
![](https://github.trello.services/images/mini-trello-icon.png) [tlvList refactoring](https://trello.com/c/JHrwp6IN/19-tlvlist-refactoring)

CoralMalachi commented on common/beerocks/tlvf/src/include/beerocks/tlvf/beerocks_header.h:
```diff
@@ -0,0 +1,66 @@
+        m_finalized = true;
+        return true;
+    }
+};
```
I did not notice this before, but since this method overrides the original finalize method, shouldn't we add here the post init check also?

tomereli replied:
You are correct. Please add a fix up commit to fix this.

CoralMalachi commented on framework/tlvf/src/include/tlvf/BaseClass.h:
```diff
@@ -38,6 +37,19 @@ class BaseClass {
+     * @param tailroom size to trim
+     * @return true on success
+     * @return false on failure
+     */
+    friend void trim(BaseClass &obj, size_t tailroom) { obj.m_buff_ptr__ -= tailroom; };
```
A friend function of a class is defined outside of a class scope, so I think we should use the friend keyword in the TlvList class, and declare the trim method as a friend method of TlvList class, so it has the right to access all private and protected members of the class and use this method.

CoralMalachi replied:
Nevermind, that is not going to work since `trim` is a private class method, but still making `trim` a friend method is wrong since it can be used by other classes, for example, CmduMessageTx, and that was not Tomer's intention.

arnout committed:
framework: tlvf: add TlvList class

Currently, the CmduMessage class conflates the CMDU-specific things with
the general concept of a list of TLVs.

Create a new class TlvList that generalizes the handling of lists of
TLVs.

In a later commit, it will be used in CmduMessage.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>
Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>
Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
tlvf: tlvlist add inner tlvlist

Preperative commit.
Add an unordered map of TLVList to the TLVlist base class which will be
used for creating/parsing of TLV internal data.

Each outer TLV will have an unknown length list for the inner TLV list
data:

outer:
	_type: class
	_is_tlv_class: true
	type: eTlvType
	length: uint16_t
	payload:
		_type: uint8_t
		_length: []

The payload is used as the buffer for the inner tlv list.
This will be used for implementing the following:
- Intel Vendor specific TLVs data - action header + action
- WSC TLVs AttrList

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: add struct TlvHeader

Preparative commit.
Add struct TlvHeader (moved from beerocks_socket_thread.h) to TlvList.h
which will be used for accessing TLVs and resising them on Finalize.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: add getTlvHeader() to TlvList

Preperative commit.
On CmduMessageTx::finalize(), we will need to call the inner TLV list
finalize, which will need to go through each TLV which contains an inner
TLV list and
1. swap it (if swap_needed = true)
2. Resize it - Since we don't know in advance the size of the inner
TLVlist, we will allocate the maximum for the inner TLV list payload,
and later will need to resize it back to its real dimensions.

Add getTlv() which will be used when resizing the TLVs on finalize.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: TlvList make finalize virtual

Preparative commit.
Make TlvList::finalize() virtual, since inner TLVs will NOT use the
default finalize since EOM TLV should only be added once.
inner TLVlist finalize would simply swap the inner TLVs data.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: BaseClass: add trim() API

Preparative commit.
Add friend function trim() which "trims" the message by tailroom length,
simply by reducing it from the current m_buff_ptr__.
This API will only be used by TLVlist on finalize, when resizing TLVs
back to their actual length.
This is a dangerous function since it doesn't do anything else but
changing the current pointer which is used for length calculations, it
should __never__ be used for anything other than this specific
operation. This is the reason it is declared private and a friend
function to TlvList alone.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: add resizeTlvs for inner TLV lists

A main TLV list may contain an inner TLV list, which is used to create /
parse a specific TLV payload.
Examples for such are the WSC TLV which contains a list of WSC
attributes, and the vendor specific TLV which contains arbitrary data
so can also be managed via an inner TLV list.

TLVs which contain inner TLV lists contain an unknown length list of
uint8_t - the payload, which is passed to the inner TLVlist constructor
as the buffer and buffer length.

When the payload is allocated, it may or may not contain variable length
data (such as variable length lists, strings, unknown length lists,
etc.), so we normally pre-allocate the payload to the maximum allowed
size. Later, on finalize(), we need to trim the tailroom of any unused
buffer space since some TLVs such as the WSC TLV does not allow to have
extranous data at the end.

This commit adds the resizeTlvs() private function to TlvList class. It
is called __only__ from the finalize() call, calculates the tailroom
of each inner TLV list, and trims the tailroom by moving the rest of the
buffer to the current TLVlist's buffer pointer.

Then it updates the tailroom of the parent TLV so the length calculation
coming afterwards is correct.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

arnout committed:
framework: tlvf: CmduMessage: use TlvList

Factor out the handling of the TLV list inside the CMDU to the TlvList
class.

It is implemented as a member, not by inheritance. This allows us to
override the functions which need to be handled differently
(getMessageLength, swap, ...) without requiring them to be virtual. On
the other hand it means we have to re-declare all the TLV handling
functions (addClass, ...) and forward to the TLV list. This can be
cleaned up later by using the tlv list directly from the callers of
these functions.

---
NOTE this implemntation is not complete yet - it builds, but is not
guarantieed to pass tests.

NOTE some of the functions are probably not needed anymore, or can be
implemented in another way.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>
Co-authored-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>
Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
common: tlvf: add beerocks_header class

Preperative commit.
Add beerocks_header header file which implements beerocks_header class
and derives from TlvList.
This will be used to replace the current beerocks vendor specific
handling and align it to use the inner TLVlists approach.

beerocks_header is not the best name for this class, since it is basically
a list of TLV like classes - first being the action header (cACTION_HEADER),
and the second is the operation itself.

A more appropriate name would have been beerocks_actions, but to keep
backward compatability, the name is kept (a comment is added in the code
in case we'll ever want to change the name.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>
Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
tlvf_conf.yaml: Remove include_source_path

Commit 5971ee4 removed copyIncludeSource() from tlvf.py.
Now that this is not used, also remove it from the tlvf_conf.yaml, and
remove it completely from tlvf.py.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

arnout committed:
tlvf: add payload to vendor specific messages

Preparative commit.
Add payload to vendor specific messages, which will be used as the
buffer for the inner TLV list - beerocks_header.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>
Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
common: beerocks_message: use beerocks_header

Preparative commit.
Refactor add_intel_vs_data() to use the VS TLV payload and create a
beerock_header class on that buffer (reminder - beerocks_header derives
from TlvList).

The payload added is the maximum allowed TLV size which is MTU - CMDU
header length - EOM TLV length. On finalize of the whole CMDU, it will
resize the buffer to its real dimensions.

Also add get_beerocks_header which in followup commits will replace
get_vs_class.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
common: beerocks_message: use get_beerocks_header

Replace get_vs_class_header with get_beerocks_header.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
common: refactor parse_intel_vs_message

Refactor parse_intel_vs_message() to return beerocks_header, and align
all callers to the new return value.

Remove cmdu_rx from all APIs which now get beerocks_header since it
shouldn't be used.

While we're at it, make all Intel VS messages use
beerocks_header->addClass instead of cmdu_rx.addClass - using mechanical
substitution:

git grep -l 'cmdu_rx\.addClass<beerocks_message' | \
  xargs sed -i 's/cmdu_rx\.addClass<beerocks_message/beerocks_header->addClass<beerocks_message/g'

This replaced __most__ occurnces but not all (due to clang formatting
issues which pushed the '.addClass' to the next line), so also manually
update those.

This is a big change, and it is not possible to split it easilly but
most of it is mechanical and results from the API changes - this was
done intentionally since we wanted compilation to break so we'll replace
all the old usages of cmdu_rx.addClass with beerocks_header->addClass.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
common: Align verify_cmdu to VS TLV changes

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: Add "jumbo" CMDUs UDS support

Since we now always preallocate the maximum tlv size for each vendor
specific TLV payload, some message passing stopped working due to the
new size restriction (even that we allow the maximum, it is not enough
in some cases. For example, the network_map smallest node is already
bigger than the MTU).

These "jumbo" messages are normally used only for internal
communication, so they do need need to be bound by the size restriction.
To allow sending "jumbo" CMDUs over UDS (internal communication), add
max_tlv_length member to CmduMessage, which can be overriden when
sending "jumbo" CMDUs.

In addition, add a wrapper to CmduMessageTx - create_vs_jumbo_message
which will first set and unset the max_tlv_length and can be used to
send jumbo CMDUs.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
network map: use create_jumbo_vs_message()

The network map is only sending messages over UDS to BML, and therefor
can send oversized TLVs using the create_jumbo_vs_message() API.
Replace all occurences of create_vs_message with create_jumbo_vs_message
using editor search-replace.
This fixes the network map which got broken due to the TLVlist refactor.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: Delete dynamicCast API from TlvList

Cfr https://github.com/prplfoundation/prplMesh/pull/551#discussion_r356575195

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: cleanups: remove getClass

Cfr https://github.com/prplfoundation/prplMesh/pull/551#discussion_r356576428

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
common: remove (now) useless template keyword

Cfr https://github.com/prplfoundation/prplMesh/pull/551#discussion_r356599019

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
remove tlvf & btlvf src from .clang-ignore

Commit 07c29131 removed the copying of source files in tlvf and btlvf
and added them as regular source files to the CMakeLists.txt.

Now that this is behind us, we no longer need to ignore these files in
clang-format, so remove these folders from .clang-ignore.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

arnout commented on common/beerocks/bcl/source/beerocks_socket_thread.cpp:
```diff
@@ -17,19 +17,15 @@
     : thread_base(), cmdu_tx(TX_BUFFER_UDS, TX_BUFFER_UDS_SIZE),
-      unix_socket_path(unix_socket_path_), server_socket(nullptr),
-      server_max_connections(DEFAULT_MAX_SOCKET_CONNECTIONS)
+      unix_socket_path(unix_socket_path_), cmdu_rx(rx_buffer + sizeof(message::sUdsHeader),
```
This change should be explained in the commit message (why it is needed, why it worked without initialisation before). Possibly in a separate commit.

tomereli replied:
Before, m_buff and m_buff_length were always set in create() and parse(). This is just silly, it makes much more sense to save the buffer and length in the constructor. I will update the relevant commit message.

arnout commented on common/beerocks/tlvf/src/include/beerocks/tlvf/beerocks_header.h:
```diff
@@ -0,0 +1,70 @@
+        : TlvList(buff, buff_len, parse, swap)
+    {
+    }
+    std::shared_ptr<beerocks_message::cACTION_HEADER> actionhdr()
+    {
+        return getClass<beerocks_message::cACTION_HEADER>();
+    }
+    beerocks_message::eAction action()
+    {
+        return actionhdr() ? actionhdr()->action() : beerocks_message::ACTION_NONE;
+    }
+    uint8_t action_op() { return actionhdr() ? actionhdr()->action_op() : 0; };
+    uint16_t id() { return actionhdr() ? actionhdr()->id() : 0; }
```
All these methods should be const.

arnout commented on common/beerocks/tlvf/src/include/beerocks/tlvf/beerocks_header.h:
```diff
@@ -0,0 +1,70 @@
+
+#include <beerocks/tlvf/beerocks_message_header.h>
+#include <tlvf/TlvList.h>
+#include <tlvf/tlvflogging.h>
```
Is this header used?

tomereli replied:
Perhaps not, will delete it if not.

arnout commented on common/beerocks/tlvf/src/include/beerocks/tlvf/beerocks_message.h:
```diff
@@ -47,46 +47,49 @@ class message_com {
-
-    static std::shared_ptr<beerocks_message::cACTION_HEADER>
-    get_vs_class_header(ieee1905_1::CmduMessage &cmdu)
+    static std::shared_ptr<beerocks_header> get_beerocks_header(ieee1905_1::CmduMessage &cmdu)
```
This get_beerocks_header() function smells funny to me...

The pattern seems to be that you do an `add_vs_tlv` which creates both the beerocks_header and the specific action object, but only returns the specific action object. Maybe it would be easier to directly return a `std::pair<std::shared_ptr<beerocks_header>, std::shared_ptr<T>>`?

It probably doesn't cover all the cases, but I feel it's a more sensible approach.

Another possibility is to go for a completely different approach. A beerocks TLV is always a VS TLV with a beerocks_action and a T. So we could have a beerocks_vs_tlv template class that has these two as members which can be accessed directly. And on the receive side we could still use beerocks_header as it is now (for receive it works fine).

(nitpick probably, or something that can be refactored later)

tomereli replied:
The approach I took here was to change as little as possible the current code, I didn't refractor the VS APIs although there surely are much better ways to do what they do. I think we can consider refactoring these as well, but in a separate PR.

arnout commented on common/beerocks/tlvf/src/include/beerocks/tlvf/beerocks_message.h:
```diff
@@ -47,46 +47,49 @@ class message_com {
     template <class T>
-    static std::shared_ptr<T> add_intel_vs_data(ieee1905_1::CmduMessageTx &cmdu_tx,
-                std::shared_ptr<ieee1905_1::tlvVendorSpecific> vs_tlv, uint16_t id = 0)
+    static bool add_intel_vs_data(ieee1905_1::CmduMessageTx &cmdu_tx, uint16_t id = 0)
```
I think it would be better to inline this directly in add_vs_tlv, so you can use the return values of the addClass calls instead of doing a getClass (and checking return value *again*) afterwards.

tomereli replied:
Or just return `std::shared_ptr<beerocks_header>`

arnout commented on common/beerocks/tlvf/src/include/beerocks/tlvf/beerocks_header.h:
```diff
@@ -0,0 +1,70 @@
+    std::shared_ptr<beerocks_message::cACTION_HEADER> actionhdr()
+    {
+        return getClass<beerocks_message::cACTION_HEADER>();
+    }
```
I don't know if it's worth it, but we could instead define

```
beerocks_message::cACTION_HEADER *operator->()
{
    auto header = getClass<beerocks_message::cACTION_HEADER();
    if (header)
        return header;
    else
        return nullptr;
}
```

arnout commented on common/beerocks/tlvf/src/include/beerocks/tlvf/beerocks_message.h:
```diff
@@ -103,27 +106,58 @@ class message_com {
+     * @return std::shared_ptr<T> to the created operation class
+     */
+    template <class T>
+    static std::shared_ptr<T> create_jumbo_vs_message(ieee1905_1::CmduMessageTx &cmdu_tx,
```
I don't like that we need to add all this special-case code (this function, max_tlv_length() and all related stuff) for something that should be rather exceptional (an internal VS message that is larger than 1450-ish bytes). But OK, I guess there's a really good reason to need this (I first thought that it was the conn_map, but that one already splits over several VS messages so it has to be something else).

tomereli replied:
I don't like it either but it's the most generic method I could think of to allow that. I reviewed the network map and it is a much bigger change to refractor that into sending smaller chunks.
If you think of a nicer way I'd love to hear it.

arnout commented on common/beerocks/tlvf/src/include/beerocks/tlvf/beerocks_message.h:
```diff
@@ -103,27 +106,58 @@ class message_com {
+        hdr->actionhdr()->action()    = (beerocks_message::eAction)action;
+        hdr->actionhdr()->action_op() = action_op;
+        hdr->actionhdr()->id()        = id;
+        cmdu_tx.tlvs.addInnerTlvList(ieee1905_1::eTlvType::TLV_VENDOR_SPECIFIC, hdr);
```
(I briefly discussed this orally with @tomereli)

What this call essentially does is to register a finalization callback - the inner tlv list is only really used in finalize (it is also used by get_beerocks_header but that's just a shorthand for passing through `getClass<ieee1905_1::eTlvType::TLV_VENDOR_SPECIFIC>`).

So maybe we should make this more explicit and instead call it setFinalizeCallback with a callable as argument. The finalize callback can be called in addClass before actually adding the class, so the resizing can be done then. Unfortunately, a showstopper for this is the swap_needed argument, which we don't know yet in addClass...

tomereli replied:
Perhaps we can just set swap_needed=True always, no matter for which message, then here we'll assume that swap_needed is true.

arnout commented on framework/tlvf/src/include/tlvf/TlvList.h:
```diff
@@ -0,0 +1,182 @@
+
+namespace ieee1905_1 {
+
+typedef struct TlvHeader {
+    uint8_t type;
+    uint16_t length;
+} __attribute__((packed)) TlvHeader;
```
Not good. TlvList should be a generic class that can be used for a list of ieee1905_1 TLVs but also for a list of WSC attributes. You can't make assumptions on the layout of the header. Unless you make it a template class which takes the header as template argument.

tomereli replied:
I think we needn't be too generic - outer TLVs, which are the only TLV types represented by a TlvList are 1905_1 TLVs - I don't see any future use for any other type. Inner TLVs are different, for those we don't need and don't use the TLV Header.

arnout commented on framework/tlvf/src/include/tlvf/TlvList.h:
```diff
@@ -0,0 +1,182 @@
+    bool is_swapped() const { return m_swapped; };
+    virtual bool finalize(bool swap_needed);
+    void reset(bool parse, bool swap);
+    void addInnerTlvList(eTlvType type, std::shared_ptr<TlvList> list)
```
Not good, we can't rely on eTlvType here because it's specific for ieee1905_1.

tomereli replied:
As I commented before, IMO outer TLVs are 1905 TLVs so we can rely on eTlvType.

arnout commented on framework/tlvf/src/src/CmduMessage.cpp:
```diff
@@ -10,149 +10,76 @@
-    } else {
-        return nullptr;
-    }
+    return tlvs.getClass<cCmduHeader>();
```
No this can't be right. The CMDU header can't be part of the list of TLVs, it's not a TLV. It doesn't start with a type and length.

tomereli replied:
IMHO A TLV list is just a wrapper for TLVF classes which do not necessarily need to be TLVs. With this assumption, and the fact that an outer TLV list is only used for 1905 TLVs we can decide that the first element is always the CMDU header. I think it makes the code simpler.

tomereli replied:
beerocks_header derives from TlvList and its classes also don't have type and length. Maybe a better name would be TlvfClassList since it's a container for TLVF classes more than it is a TLV list. Then if we really want to, we can always extend it and create a strict TLV only list.

arnout replied:
> beerocks_header derives from TlvList and its classes also don't have type and length. 

Good point, I didn't realize that. I was too fixated on WSC :-)

> Maybe a better name would be TlvfClassList since it's a container for TLVF classes more than it is a TLV list.

That's a different idea than what I originally had, but makes a lot more sense. Indeed, if you look at it as a sequence of tlvf-generated structures, it makes a lot more sense.

So yes, a rename would be a good idea. ClassList is maybe more appropriate, like we have BaseClass.

arnout commented on framework/tlvf/src/include/tlvf/BaseClass.h:
```diff
@@ -38,6 +37,19 @@ class BaseClass {
+     * @return true on success
+     * @return false on failure
+     */
+    friend void trim(BaseClass &obj, size_t tailroom) { obj.m_buff_ptr__ -= tailroom; };
```
I think the trim is not appropriate here. I think it would make more sense to make a special case in tlvf for when an unknown-length list is at the end of the class (or maybe even make a special _payload type for this). In that specific case, we could add two functions: a prepare that reserves the rest of the buffer and returns the remaining length, and a commit with the final length as argument that updates the end pointer and the length field. Both functions are pretty much the same as allocate (the trim is like allocate with a negative value) in that they update both length and end pointer.

tomereli replied:
Sound OK, but the commit API doesn't need an argument, it knows the buffer length and we know the current buffer pointer so we can calculate that ourselves

arnout reviewed (changes requested):
I did a "quick" review to give feedback if I thought this was going in the right direction. Unfortunately, the answer is "not entirely". I'm not convinced of the solution for updating the length of inner tlv lists, it feels too complicated and too error-prone to me. That said, we *may* decide to merge it like this anyway and refactor later.

To make some progress, I would like a separate PR with the following commits:

- 41d99bcc tlvf: Fix bug in BuffPtrIncrementSafe()
- 4531edde remove tlvf & btlvf src from .clang-ignore
- e80a2ce1 tlvf_conf.yaml: Remove include_source_path (note that the commit sha referenced in the commit message is not correct)
- cda27dbe common: beerocks_message: remove unclear TODO
- b7792fa6 tlvf: Delete dynamicCast API from TlvList


And maybe a second PR with:

- f5bacbbf framework: tlvf: add TlvList class
- 3a1c3f47 tlvf: add payload to vendor specific messages

tomereli commented:
> 
> To make some progress, I would like a separate PR with the following commits:
> 
> * [41d99bc](https://github.com/prplfoundation/prplMesh/commit/41d99bcc3d6f26e47392b6a6f12dd437602bc3b3) tlvf: Fix bug in BuffPtrIncrementSafe()
> * [4531edd](https://github.com/prplfoundation/prplMesh/commit/4531edde8dd65be85c8fa3fe82d27c779e783490) remove tlvf & btlvf src from .clang-ignore
> * [e80a2ce](https://github.com/prplfoundation/prplMesh/commit/e80a2ce1a48e67ed8003d4e3d458707f1fb5fe7f) tlvf_conf.yaml: Remove include_source_path (note that the commit sha referenced in the commit message is not correct)
> * [cda27db](https://github.com/prplfoundation/prplMesh/commit/cda27dbe263978a7599a038ad703edcfd47040a6) common: beerocks_message: remove unclear TODO

This

> * [b7792fa](https://github.com/prplfoundation/prplMesh/commit/b7792fa6dc2277202406debd6f0486bb99a83934) tlvf: Delete dynamicCast API from TlvList

can't be part of this PR since TlvList is not yet added.

Commit:
fix after rebase

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
common: socket thread: always swap on finalize

Preparative commit.

Always swap on finalize - we want to replace the finalize call so that
each BaseClass will implement its own finalize and will do the swap
there. The plan is to update addClass to always call finalize for the
previous class which was added. This will do the swapping always (we
don't have swap_needed in addClass), and will do the resizing if inner
TLVlist is present for that TLV - so the TLV finalize will call the
inner TLV finalize which will call finalize for every TLV in the inner
TLV vector and so forth.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tree-wide: completely remove m_swap and swap_needed

From all the non-autogenerated files.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: BaseClass: remove m_swap completely

Now that we __always__ swap, remove the unused m_swap member from
BaseClass and update all callers. Also update tlvf.py for all the
autogenerated files, but don't update them as part of this commit, they
will be updated in a separate commit to make cherry-picking to #584
easier and also because we never want such a huge change in one commit
to hide the non autogen changes.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: update autogen files after m_swap removal

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>
